benches/dapol.rs

```diff
@@ -1,175 +1,250 @@
+        1
+    } else {
+        max_thread_count >> 2
+    };
 
-    group.finish();
-}
+    e.advance().unwrap();
 
-fn verify_proof(c: &mut Criterion) {
-    let mut group = c.benchmark_group("verify");
-    group.sample_size(10);
+    for i in (step..max_thread_count).step_by(step as usize) {
+        e.advance().unwrap();
+        thread_counts.push(i);
+    }
 
-    // this benchmark depends on the tree height and not the number of leaves,
-    // so we just pick the smallest number of leaves
-    let num_leaves = NUM_LEAVES[0];
-    for &tree_height in TREE_HEIGHTS.iter() {
-        let items = build_item_list(num_leaves, tree_height);
-        let mut rng = thread_rng();
-        let item_range = Uniform::new(0usize, num_leaves);
-
-        let dapol = build_dapol_tree::<blake3::Hasher, RangeProofSplitting>(&items, tree_height);
-        group.bench_function(BenchmarkId::new("splitting", tree_height), |bench| {
-            bench.iter_batched(
-                || {
-                    // generate a proof
-                    let item_idx = rng.sample(item_range);
-                    let tree_index = &items[item_idx].0;
-                    (item_idx, dapol.generate_proof(tree_index).unwrap())
-                },
-                |(item_idx, proof)| {
-                    // time proof verification
-                    proof.verify(&dapol.root(), &items[item_idx].1.get_proof_node())
-                },
-                BatchSize::SmallInput,
-            );
-        });
-
-        let dapol = build_dapol_tree::<blake3::Hasher, RangeProofPadding>(&items, tree_height);
-        group.bench_function(BenchmarkId::new("padding", tree_height), |bench| {
-            bench.iter_batched(
-                || {
-                    // generate a proof
-                    let item_idx = rng.sample(item_range);
-                    let tree_index = &items[item_idx].0;
-                    (item_idx, dapol.generate_proof(tree_index).unwrap())
-                },
-                |(item_idx, proof)| {
-                    // time proof verification
-                    proof.verify(&dapol.root(), &items[item_idx].1.get_proof_node())
-                },
-                BatchSize::SmallInput,
-            );
-        });
+    thread_counts.push(max_thread_count);
+
+    let mut ndm_smt = Option::<NdmSmt>::None;
+
+    e.advance().unwrap();
+
+    for h in TREE_HEIGHTS.into_iter() {
+        e.advance().unwrap();
+
+        for t in thread_counts.iter() {
+            e.advance().unwrap();
+
+            for u in NUM_USERS.into_iter() {
+                e.advance().unwrap();
+
+                let max_users_for_height = 2_u64.pow((h - 1) as u32);
+
+                if u > max_users_for_height {
+                    break;
+                }
+
+                let tup: (Height, MaxThreadCount, u64) =
+                    (Height::from(h), MaxThreadCount::from(*t), u);
+
+                // tree build compute time
+                group.bench_function(
```

Comment:
What's the reason for using `bench_function` over `bench_with_input`? Have you read [this](https://bheisler.github.io/criterion.rs/book/user_guide/benchmarking_with_inputs.html)? It seems better suited to our needs. You can also get nice graphs it seems. The example [here](https://docs.rs/criterion/0.5.1/criterion/struct.BenchmarkGroup.html#examples) is very similar to our benches.